### PR TITLE
Pass filters argument to compile funcs and PreProcessor

### PR DESCRIPTION
**Pull request by moyogo**
Continues #466 and is needed for https://github.com/googlefonts/fontmake/pull/737.

This also adds BaseFilter.pre to specify if the filter is a pre-filter or a post-filter. It is False, for post-filter, by default, as `loadFilter()` puts filters in postFilters by default.

### CODE REVIEW

File: Lib/ufo2ft/filters/__init__.py [hunk 1]
```diff
@@ -68,10 +68,11 @@ def loadFilters(ufo):
         filterObj = filterClass(
             include=filterDict.get("include"),
             exclude=filterDict.get("exclude"),
+            pre=filterDict.get("pre"),
```
Comment (anthrotype) [inline Lib/ufo2ft/filters/__init__.py:71]:
```suggestion
            pre=filterDict.get("pre", False),
```

File: Lib/ufo2ft/preProcessor.py [hunk 2]
```diff
@@ -242,11 +253,14 @@ def __init__(
...
-            self.preFilters.append(pre)
-            self.postFilters.append(post)
+        if filters is None:
+            self.preFilters, self.postFilters = [], []
+            for ufo in ufos:
+                pre, post = loadFilters(ufo)
+                self.preFilters.append(pre)
+                self.postFilters.append(post)
+        else:
+            self.preFilters, self.postFilters = ((f,) * len(ufos) for f in filters)
```
Comment (anthrotype) [inline Lib/ufo2ft/preProcessor.py:263]: this is incorrect, you're not splitting the pre and post filters.

**Review by anthrotype — COMMENTED** [review]
Thanks, I left some comments. Also, we'd need tests for these new feature

File: tests/preProcessor_test.py [hunk 3]
```diff
@@ -77,6 +77,45 @@ def test_inplace_no_remember_curve_type(self, FontClass):
...
+        glyphSets0 = TTFPreProcessor(ufo1).process()
+        glyphSets1 = TTFPreProcessor(ufo2).process()
+
+        assert (glyphSets0["a"][0][0].x - glyphSets1["a"][0][0].x) == -40
+        assert (glyphSets1["a"][0][0].y - glyphSets0["a"][0][0].y) == 10
+
+    def test_custom_filters_as_argument(self, FontClass):
+        ufo1 = FontClass(getpath("TestFont.ufo"))
+        ufo2 = FontClass(getpath("TestFont.ufo"))
+        filter1 = loadFilterFromString("RemoveOverlapsFilter(backend='pathops')")
```
Comment (anthrotype) [inline tests/preProcessor_test.py:97]: why loading from string? There ought to be separate unit tests for that, here you're only testing that you can pass a `filters` argument

**Review by anthrotype — COMMENTED** [review]
maybe add another test in integration_test.py testing that the filters parameter is actually passed through to the preprocessor from all the compile* call sites

**Comment by moyogo** [comment]
> maybe add another test in integration_test.py testing that the filters parameter is actually passed through to the preprocessor from all the compile* call sites

Good catch, some filters args were missing.

File: tests/integration_test.py [hunk 4]
```diff
@@ -251,6 +253,81 @@ def test_drop_glyph_names_variable(
         ttf = compile_func(designspace, **options)
         expectTTX(ttf, expected_ttx)
 
+    @pytest.mark.parametrize(
+        "compileFunc",
+        [
+            (compileOTF),
+            (compileTTF),
```
Comment (anthrotype) [inline tests/integration_test.py:260]: these () parentheses don't do anything, you can remove them

File: tests/integration_test.py [hunk 5]
```diff
@@ -251,6 +253,81 @@ def test_drop_glyph_names_variable(
...
+        assert pen1.bounds[0] == pen2.bounds[0]
+        assert pen1.bounds[1] + 10 == pen2.bounds[1]
+        assert pen1.bounds[2] == pen2.bounds[2]
+        assert pen1.bounds[3] + 10 == pen2.bounds[3]
+
+    @pytest.mark.parametrize(
+        "compileFunc",
+        [
+            (compileVariableTTF),
+            (compileVariableCFF2),
```
Comment (anthrotype) [inline tests/integration_test.py:286]: same here, remove ()

**Review by anthrotype — APPROVED** [review]

**Pull request merged**